### PR TITLE
[SPARK-36895][SQL] Add Create Index syntax support

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
This is the 2nd PR for DSv2 index support.

This PR adds the following:

- create index syntax support in parser and analyzer
- `CreateIndex` logic node
- `CreateIndexExec` physical node

`CreateIndex` is not implemented yet in this PR. Calling `CreateIndex` will throw `SQLFeatureNotSupportedException`, and the parsed index information such as `IndexName` `indexType` `columns` and index properties will be included in the error message for now for testing purpose.

### Why are the changes needed?
To support index in DSv2

### Does this PR introduce _any_ user-facing change?
Yes, the create table syntax as the following:

```
CREATE INDEX index_name ON [TABLE] table_name [USING index_type] (column_index_property_list)[OPTIONS indexPropertyList]

    column_index_property_list: column_name [OPTIONS(indexPropertyList)]  [ ,  . . . ]
    indexPropertyList: index_property_name [= index_property_value] [ ,  . . . ]
```

### How was this patch tested?
add a UT

### CODE REVIEW

**Comment by huaxingao** [comment]
@cloud-fan @viirya 
Could you please take a look?
`SQLKeywordSuite` failed because the new keywords `BLOOM_FILTER_INDEX`, `BTREE_INDEX` and `Z_ORDERING_INDEX`  for index type are not in the documentation yet.

**Comment by huaxingao** [comment]
I removed these specific index types `BLOOM_FILTER_INDEX`, `BTREE_INDEX` and `Z_ORDERING_INDEX` for now to pass the `SQLKeywordSuite`. We can discuss what index types to support and document in next PR after Wenchen comes back from the holiday break.

**Comment by huaxingao** [comment]
@dongjoon-hyun @sunchao @viirya 
This PR is ready for review. The failed test `MariaDBKrbIntegrationSuite` is not relevant.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala [hunk 1]
```diff
@@ -1022,3 +1022,15 @@ case class UncacheTable(
...
 }
+
+/**
+ * The logical plan of the CREATE INDEX command.
+ */
+case class CreateIndex(
+    child: LogicalPlan,
+    indexName: String,
+    indexType: String,
+    ifNotExists: Boolean,
```
Comment (huaxingao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala:1033]: This should be `ignoreIfExists`. I will fix this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * Physical plan node for creating an index.
+ */
+case class CreateIndexExec(
+    catalog: TableCatalog,
+    ident: Identifier,
+    indexName: String,
+    indexType: String,
+    ifNotExists: Boolean,
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala:36]: This should be `ignoreIfExists`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -4492,6 +4502,55 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+      columnsProperties,
+      options)
+  }
+
+  /**
+   * Convert a index property list into a key-value map.
+   */
+  override def visitIndexPropertyList(
+      ctx: IndexPropertyListContext): Map[String, String] = withOrigin(ctx) {
+    val temp = ctx.property
```
Comment (huaxingao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4541]: will remove

**Comment by HyukjinKwon** [comment]
@huaxingao just a quick question. do you have an implementation of the index support in DSv2? I am asking this because pandas API on Spark already implemented this actually. I would like to investigate the feasibility of migration in the future more just for curiosity .. If you don't have it now, it should be good to have one in test or documentation.

**Comment by huaxingao** [comment]
@HyukjinKwon Thanks for taking a look! I am implementing this in JDBC right now, so I can test out the new code in index support.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala [hunk 4]
```diff
@@ -425,4 +425,27 @@ class JDBCTableCatalogSuite extends QueryTest with SharedSparkSession {
...
+    withTable("h2.test.new_table") {
+      sql("CREATE TABLE h2.test.new_table(col1 INT, col2 STRING)")
+      val e1 = intercept[SQLFeatureNotSupportedException] {
+        sql("CREATE index i1 ON h2.test.new_table (col1)")
+      }.getMessage
+      assert(e1.contains("IndexName i1"))
+      assert(e1.contains("indexType "))
+      assert(e1.contains("columns ArrayBuffer(col1)"))
+
+      val e2 = intercept[SQLFeatureNotSupportedException] {
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala:439]: can we test more combinations? e.g., when there is no option given to a column, or no option is given to the index, or an index type is specified etc.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 5]
```diff
@@ -110,9 +110,9 @@ statement
     | CREATE namespace (IF NOT EXISTS)? multipartIdentifier
         (commentSpec |
          locationSpec |
-         (WITH (DBPROPERTIES | PROPERTIES) tablePropertyList))*        #createNamespace
+         (WITH (DBPROPERTIES | PROPERTIES) propertyList))*             #createNamespace
```
Comment (viirya) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:113]: Seems you add some additional spaces?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 6]
```diff
@@ -2785,11 +2785,11 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
    * Convert a table property list into a key-value map.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:2785]: not `table property` anymore.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 7]
```diff
@@ -4492,6 +4492,37 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
   }
 
+  /**
+   * Create an index, returning a [[CreateIndex]] logical plan.
+   * For example:
+   * {{{
+   * CREATE [index_type] INDEX [index_name] ON [TABLE] table_name (column_index_property_list)
+   *   [OPTIONS indexPropertyList]
+   *   column_index_property_list: column_name [OPTIONS(indexPropertyList)]  [ ,  . . . ]
+   *   indexPropertyList: index_property_name = index_property_value [ ,  . . . ]
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4502]: Hmm, is `index_property_value` required or optional?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 8]
```diff
@@ -4492,6 +4492,37 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+    val columns = ctx.columns.multipartIdentifierProperty.asScala
+      .map(_.multipartIdentifier.getText).toSeq
+    val columnsProperties = ctx.columns.multipartIdentifierProperty.asScala
+      .map(x => (Option(x.options).map(visitPropertyKeyValues).getOrElse(Map.empty))).toSeq
+    val options = Option(ctx.options).map(visitPropertyKeyValues).getOrElse(Map.empty)
+
+    CreateIndex(
+      UnresolvedDBObjectName(
+        visitMultipartIdentifier(ctx.multipartIdentifier),
+        isNamespace = true),
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4517]: We only can create index on a table, right? If so, this sounds to be `UnresolvedTable` and we can use `createUnresolvedTable(...)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala [hunk 9]
```diff
@@ -427,6 +427,11 @@ class DataSourceV2Strategy(session: SparkSession) extends Strategy with Predicat
       val table = a.table.asInstanceOf[ResolvedTable]
       AlterTableExec(table.catalog, table.identifier, a.changes) :: Nil
 
+    case CreateIndex(ResolvedDBObjectName(catalog, table),
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:430]: ditto. I think we cannot create index on arbitrary database object.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 7 above)
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4502]: It's optional. I forgot to put this in `[]`.
I double checked oracle and MySQL reference, the index properties could have name without value.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 8 above)
Comment (huaxingao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4517]: You are right. Should be `UnresolvedTable`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 10]
```diff
@@ -4492,6 +4492,35 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
     CommentOnTable(createUnresolvedTable(ctx.multipartIdentifier, "COMMENT ON TABLE"), comment)
   }
 
+  /**
+   * Create an index, returning a [[CreateIndex]] logical plan.
+   * For example:
+   * {{{
+   * CREATE [index_type] INDEX [index_name] ON [TABLE] table_name (column_index_property_list)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4499]:
```suggestion
   * CREATE [index_type] INDEX index_name ON [TABLE] table_name (column_index_property_list)
```

index name is not optional

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    indexType: String,
+    ignoreIfExists: Boolean,
+    columns: Seq[FieldReference],
+    columnProperties: Seq[Map[String, String]],
+    private var properties: Map[String, String])
+  extends LeafV2CommandExec {
+  override protected def run(): Seq[InternalRow] = {
+
+    try {
+      // Todo: replace this with the createIndex
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala:44]: The actual implementation is as simple as calling v2 `createIndex` API, right? I think we should throw exception in the testing v2 source instead.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala [hunk 12]
```diff
@@ -425,4 +425,45 @@ class JDBCTableCatalogSuite extends QueryTest with SharedSparkSession {
       assert(m.contains("\"TABLEENGINENAME\" not found"))
     }
   }
+
+  test("CREATE INDEX") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCTableCatalogSuite.scala:429]: can we also add parser UT?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 13]
```diff
@@ -244,6 +244,9 @@ statement
     | SET .*?                                                          #setConfiguration
     | RESET configKey                                                  #resetQuotedConfiguration
     | RESET .*?                                                        #resetConfiguration
+    | CREATE indexType=STRING? INDEX (IF NOT EXISTS)? identifier ON TABLE?
```
Comment (cloud-fan) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:247]:
```suggestion
    | CREATE indexType=identifier? INDEX (IF NOT EXISTS)? identifier ON TABLE?
```

STRING only matches `'abc'`, but we want to support `CREATE bloom_filter INDEX ...`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala (hunk 11 above)
Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/CreateIndexExec.scala:44]: I have JDBC v2 source `createIndex` implemented in https://github.com/apache/spark/pull/34164. We can probably merge that PR first, and then I can remove this `Todo`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 14]
```diff
@@ -4492,6 +4492,39 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+   * {{{
+   * CREATE [index_type] INDEX index_name ON [TABLE] table_name (column_index_property_list)
+   *   [OPTIONS indexPropertyList]
+   *   column_index_property_list: column_name [OPTIONS(indexPropertyList)]  [ ,  . . . ]
+   *   indexPropertyList: index_property_name [= index_property_value] [ ,  . . . ]
+   * }}}
+   */
+  override def visitCreateIndex(ctx: CreateIndexContext): LogicalPlan = withOrigin(ctx) {
+    val (indexType, indexName) = if (ctx.identifier.size() == 1) {
+      (null, ctx.identifier(0).getText)
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:4507]: Sorry just for my own understanding, why do we allow creating index without index type?

**Comment by huaxingao** [comment]
@cloud-fan I have changed `CREATE [index_type] INDEX ...` to `CREATE INDEX index_name ON [TABLE] table_name [USING index_type]`. Please take a look again when you have time. Thanks a lot!

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Pull request closed without merging**